### ISSUE TITLE
get_weights_as_images

### ISSUE THREAD

**Issue by WillieMaddox** [blackecho/Deep-Learning-TensorFlow#25]
I think the first argument to the function "yadlt/utils/utilities/get_weights_as_images" should not be "self"?  Correct me if I'm wrong.  Also, could you provide an example of how to use this function.  It seems pretty straight forward except for the "weights" argument, not sure what to pass here.  Can we simply use one of the .npy output files from your examples (e.g. one of the output layers from command_line/run_stacked_autoencoder_unsupervised.py)?

**Comment by blackecho** [comment]
Yes, the "self" argument shouldn't be there. I've just removed it. I also changed the first argument from weights to weights_npy. Now you can simply pass the path to the weights .npy file and it will be loaded. weights is expected to be a 2-dimensional matrix, so you can pass only the weights of a single layer.

**Comment by WillieMaddox** [comment]
Awesome!  In your code examples, are the weights just the output layers?

**Comment by blackecho** [comment]
The output layers are the activations of the layers when the dataset is fed to the network.
The weights of the model in stacked autoencoders are in the properties `self.encoding_w`, `self.encoding_b`, `self.decoding_w` and `self.decoding_b`.
If you're using run_stacked_autoencoder_unsupervised.py from command_line it doesn't have this option yet. I'll add it asap.

**Comment by blackecho** [comment]
Just added the option. Just run command_line/run_stacked_autoencoder_unsupervised.py with the option `--save_model_parameters /path/to/directory` and the numpy arrays of the weights and biases will be saved to that directory.

**Comment by WillieMaddox** [comment]
I was able to obtain weights using the "save_model_parameters" flag on the mnist dataset for the stacked deep autoencoder.  But I am having issues displaying the kernels using the "get_weights_as_images" function.

For example, using the code provided in the sdae example, the shape of resulting "enc-weights-layer-3.npy" is (512, 784).  But, the square root of 512 is 22.627, so I'm not sure what to put in for width and height:

```python
utilities.get_weights_as_images(os.path.join(datapath, "dec-weights-layer-3.npy"), ???, ???)
```

If I modify the get_weights_as_images function to take the transpose of the weights and set width and height to 28, that gets you past the "len(img) == width * height" assertion error, but the resulting kernel images just look like noisy garbage.

So my question is what is the correct way to use the "get_weights_as_images" function to view these "save_model_parameters" weights?

I can provide more details if you need them.

Thanks again.

**Comment by blackecho** [comment]
Unfortunately, that simple function can be used only to visualize the kernels of the first hidden layer, for which width and height are known (28x28 in the case of MNIST) and there is a direct relation with the input.
If you want to visualize the kernels of higher layers, more sophisticated techniques must be used, like [this one](http://yosinski.com/deepvis). Unfortunately, I still have to implement it.

**Comment by WillieMaddox** [comment]
I also tried using the function on the first layer but I'm still getting noisy garbage.  I was expecting to get something more along the lines of [this](http://deeplearning.stanford.edu/wiki/index.php/Visualizing_a_Trained_Autoencoder).  Is this reasonable?

**Comment by blackecho** [comment]
Yes they should be like those ones. Maybe this a problem with the training hyper-parameters. The ones that I wrote in the documentation are not optimized. I obtained very good results on MNIST with this parameters:
`layers: 2048,1024,256,64`,  `encoder act func: sigmoid, sigmoid, sigmoid, sigmoid` `decoder act func: sigmoid, sigmoid, sigmoid, sigmoid`, `optimizer: adam`, `learning rate: 1e-4`, `batch_size: 64`, `num_epochs: 120` `do pretrain: false`

**Comment by WillieMaddox** [comment]
yeah, I still get noise with those parameters.  Here is the command I used (assume a space after each line):

time python command_line/run_stacked_autoencoder_unsupervised.py
--dataset mnist
--dae_layers 2048,1024,256,64
--dae_batch_size 64
--dae_num_epochs 120
--dae_learning_rate 1e-4
--dae_enc_act_func sigmoid, sigmoid, sigmoid, sigmoid
--dae_dec_act_func sigmoid, sigmoid, sigmoid, sigmoid
--dae_opt adam
--do_pretrain false
--verbose 1
--save_reconstructions
/home/maddoxw/PycharmProjects/Deep-Learning-TensorFlow/res/sdae/mnist/recon.npy
--save_layers_output_test
/home/maddoxw/PycharmProjects/Deep-Learning-TensorFlow/res/sdae/mnist
--save_layers_output_train
/home/maddoxw/PycharmProjects/Deep-Learning-TensorFlow/res/sdae/mnist
--save_model_parameters
/home/maddoxw/PycharmProjects/Deep-Learning-TensorFlow/res/sdae/mnist/

And the output:

I tensorflow/stream_executor/dso_loader.cc:108] successfully opened CUDA library libcublas.so.7.5 locally
I tensorflow/stream_executor/dso_loader.cc:108] successfully opened CUDA library libcudnn.so locally
I tensorflow/stream_executor/dso_loader.cc:108] successfully opened CUDA library libcufft.so.7.5 locally
I tensorflow/stream_executor/dso_loader.cc:108] successfully opened CUDA library libcuda.so.1 locally
I tensorflow/stream_executor/dso_loader.cc:108] successfully opened CUDA library libcurand.so.7.5 locally
Extracting MNIST_data/train-images-idx3-ubyte.gz
Extracting MNIST_data/train-labels-idx1-ubyte.gz
Extracting MNIST_data/t10k-images-idx3-ubyte.gz
Extracting MNIST_data/t10k-labels-idx1-ubyte.gz
Creating /home/maddoxw/.yadlt/stored_models/un_sdae/ directory to save/restore models
Creating /home/maddoxw/.yadlt/data/un_sdae/ directory to save model generated data
Creating /home/maddoxw/.yadlt/logs/un_sdae/ directory to save tensorboard logs
Creating /home/maddoxw/.yadlt/stored_models/un_sdae/dae-1 directory to save/restore models
Creating /home/maddoxw/.yadlt/data/un_sdae/dae-1 directory to save model generated data
Creating /home/maddoxw/.yadlt/logs/un_sdae/dae-1 directory to save tensorboard logs
Creating /home/maddoxw/.yadlt/stored_models/un_sdae/dae-2 directory to save/restore models
Creating /home/maddoxw/.yadlt/data/un_sdae/dae-2 directory to save model generated data
Creating /home/maddoxw/.yadlt/logs/un_sdae/dae-2 directory to save tensorboard logs
Creating /home/maddoxw/.yadlt/stored_models/un_sdae/dae-3 directory to save/restore models
Creating /home/maddoxw/.yadlt/data/un_sdae/dae-3 directory to save model generated data
Creating /home/maddoxw/.yadlt/logs/un_sdae/dae-3 directory to save tensorboard logs
Creating /home/maddoxw/.yadlt/stored_models/un_sdae/dae-4 directory to save/restore models
Creating /home/maddoxw/.yadlt/data/un_sdae/dae-4 directory to save model generated data
Creating /home/maddoxw/.yadlt/logs/un_sdae/dae-4 directory to save tensorboard logs
I tensorflow/core/common_runtime/gpu/gpu_init.cc:102] Found device 0 with properties: 
name: GeForce GTX TITAN X
major: 5 minor: 2 memoryClockRate (GHz) 1.2155
pciBusID 0000:03:00.0
Total memory: 12.00GiB
Free memory: 9.93GiB
I tensorflow/core/common_runtime/gpu/gpu_init.cc:126] DMA: 0 
I tensorflow/core/common_runtime/gpu/gpu_init.cc:136] 0:   Y 
I tensorflow/core/common_runtime/gpu/gpu_device.cc:843] Creating TensorFlow device (/gpu:0) -> (device: 0, name: GeForce GTX TITAN X, pci bus id: 0000:03:00.0)
Tensorboard logs dir for this run is /home/maddoxw/.yadlt/logs/un_sdae/run14
Reconstruction loss at step 0: 0.597067
Reconstruction loss at step 1: 0.398617
Reconstruction loss at step 2: 0.333239
Reconstruction loss at step 3: 0.307402
Reconstruction loss at step 4: 0.294534
Reconstruction loss at step 5: 0.287009
Reconstruction loss at step 6: 0.282103
Reconstruction loss at step 7: 0.278649
Reconstruction loss at step 8: 0.276077
Reconstruction loss at step 9: 0.274082
Reconstruction loss at step 10: 0.272482
Reconstruction loss at step 11: 0.27117
Reconstruction loss at step 12: 0.270067
Reconstruction loss at step 13: 0.269124
Reconstruction loss at step 14: 0.268306
Reconstruction loss at step 15: 0.267589
Reconstruction loss at step 16: 0.26695
Reconstruction loss at step 17: 0.266378
Reconstruction loss at step 18: 0.265859
Reconstruction loss at step 19: 0.265384
Reconstruction loss at step 20: 0.264947
Reconstruction loss at step 21: 0.264539
Reconstruction loss at step 22: 0.264156
Reconstruction loss at step 23: 0.263793
Reconstruction loss at step 24: 0.26345
Reconstruction loss at step 25: 0.263119
Reconstruction loss at step 26: 0.262802
Reconstruction loss at step 27: 0.262496
Reconstruction loss at step 28: 0.262199
Reconstruction loss at step 29: 0.261912
I tensorflow/core/common_runtime/gpu/gpu_device.cc:843] Creating TensorFlow device (/gpu:0) -> (device: 0, name: GeForce GTX TITAN X, pci bus id: 0000:03:00.0)
Test set reconstruction loss: 0.261689394712
Saving the reconstructions for the test set...
I tensorflow/core/common_runtime/gpu/gpu_device.cc:843] Creating TensorFlow device (/gpu:0) -> (device: 0, name: GeForce GTX TITAN X, pci bus id: 0000:03:00.0)
Saving the output of each layer for the test set
I tensorflow/core/common_runtime/gpu/gpu_device.cc:843] Creating TensorFlow device (/gpu:0) -> (device: 0, name: GeForce GTX TITAN X, pci bus id: 0000:03:00.0)
Saving the output of each layer for the train set
I tensorflow/core/common_runtime/gpu/gpu_device.cc:843] Creating TensorFlow device (/gpu:0) -> (device: 0, name: GeForce GTX TITAN X, pci bus id: 0000:03:00.0)
Saving the parameters of the model
I tensorflow/core/common_runtime/gpu/gpu_device.cc:843] Creating TensorFlow device (/gpu:0) -> (device: 0, name: GeForce GTX TITAN X, pci bus id: 0000:03:00.0)

real	11m35.368s
user	15m48.192s
sys	1m13.580s


Thoughts? Maybe you'll catch something I missed.

**Comment by blackecho** [comment]
The `do_pretrain` option skips the pre-training of the autoencoders and goes directly to finetuning. Thus, the `dae_*` parameters are ignored, you should modify `dae_*` with `finetune_*` (except for `dae_layers`).
In the following parameters the prefix should be `finetune_*` instead of `dae_*`
--dae_batch_size 64
--dae_num_epochs 120
--dae_learning_rate 1e-4
--dae_enc_act_func sigmoid, sigmoid, sigmoid, sigmoid
--dae_dec_act_func sigmoid, sigmoid, sigmoid, sigmoid
--dae_opt adam

**Comment by WillieMaddox** [comment]
That works much better. Thank you!
